data/binding/binding.go

```diff
@@ -0,0 +1,58 @@
+//go:generate go run gen.go
+
+package binding
+
+// DataItem is the base interface for all bindable data items.
+type DataItem interface {
+	AddListener(DataItemListener)
+	RemoveListener(DataItemListener)
+}
+
+// DataItemListener is any object that can register for changes in a bindable DataItem.
+// See NewDataItemListener to define a new listener using just an inline function.
+type DataItemListener interface {
+	DataChanged(DataItem)
+}
+
+// NewDataItemListener is a helper function that creates a new listener type from a simple callback function.
+func NewDataItemListener(fn func(DataItem)) DataItemListener {
+	return &listener{fn}
+}
+
+type listener struct {
+	callback func(DataItem)
+}
+
+func (l *listener) DataChanged(i DataItem) {
+	l.callback(i)
+}
+
+type base struct {
+	listeners []DataItemListener
+}
+
+// AddListener allows a data listener to be informed of changes to this item.
+func (b *base) AddListener(l DataItemListener) {
+	b.listeners = append(b.listeners, l)
+}
+
+// RemoveListener should be called if the listener is no longer interested in being informed of data change events.
+func (b *base) RemoveListener(l DataItemListener) {
+	for i, listen := range b.listeners {
+		if listen != l {
+			continue
+		}
+
+		if i == len(b.listeners)-1 {
+			b.listeners = b.listeners[:len(b.listeners)-1]
+		} else {
+			b.listeners = append(b.listeners[:i], b.listeners[i+1:]...)
+		}
+	}
+}
+
+func (b *base) trigger(i DataItem) {
+	for _, listen := range b.listeners {
+		listen.DataChanged(i)
```

Comment:
Do we not want to execute listeners in separate goroutines?

Reply:
This is an interesting question. I know we went round it before - but did we come to a conclusion?
It can't be new goroutines due to ordering requirements but it could be a background thread like the svg cache janitor etc.
One more consideration is whether we expect third party bindings - if so they may have to hook into the same (or at least be able to, i.e. BaseBinding) to ensure that their code does not break our design.

Reply:
Are we agreed that it cannot be as it currently is, as (some/most?) data binding changes will come from the UI ie GLFW event thread?

Reply:
If we've built things currently then no developer code can run on the main thread (GLFW events are on main).
I don't think that currently it is possible that the binding will be called from main thread.
However it is likely that it may become possible if app properties or window management makes use of this so it should probably change before that happens.

Reply:
I assumed things like `OnTapped`, `TypedKey`, and `TypedRune` were run on the event thread

Reply:
"the event thread" is our event thread not GLFW's main event thread.

Reply:
Added a data item thread